src/coprocessor/coprocessor.rs

```diff
@@ -0,0 +1,1142 @@
+            tp => return Err(box_err!("unsupported tp {}", tp)),
+        };
+        Ok((builder, req_ctx))
+    }
+
+    /// Parse the raw `Request` to create `RequestHandlerBuilder` and `ReqContext`.
+    #[inline]
+    fn parse_request(
+        &self,
+        req: coppb::Request,
+        peer: Option<String>,
+        is_streaming: bool,
+    ) -> (RequestHandlerBuilder<E::Snap>, ReqContext) {
+        match self.try_parse_request(req, peer, is_streaming) {
+            Ok(v) => v,
+            Err(err) => {
+                // If there are errors when parsing requests, create a dummy request handler.
+                let builder =
+                    box |_, _: &_| Ok(cop_util::ErrorRequestHandler::new(err).into_boxed());
+                let req_ctx = ReqContext::new(
+                    "invalid",
+                    kvrpcpb::Context::new(),
+                    &[],
+                    Duration::from_secs(60), // Large enough to avoid becoming outdated error
+                    None,
+                    None,
+                    None,
+                );
+                (builder, req_ctx)
+            }
+        }
+    }
+
+    /// Get the batch row limit configuration.
+    #[inline]
+    fn get_batch_row_limit(&self, is_streaming: bool) -> usize {
+        if is_streaming {
+            self.stream_batch_row_limit
+        } else {
+            self.batch_row_limit
+        }
+    }
+
+    #[inline]
+    fn async_snapshot(
+        engine: E,
+        ctx: &kvrpcpb::Context,
+    ) -> impl Future<Item = E::Snap, Error = Error> {
+        let (callback, future) = ::util::future::paired_future_callback();
+        let val = engine.async_snapshot(ctx, callback);
+        future::result(val)
+            .and_then(|_| future.map_err(|cancel| storage::engine::Error::Other(box_err!(cancel))))
+            .and_then(|(_ctx, result)| result)
+            // map engine::Error -> coprocessor::Error
+            .map_err(Error::from)
+    }
+
+    /// The real implementation of handling a unary request.
+    ///
+    /// It first retrieves a snapshot, then builds the `RequestHandler` over the snapshot and
+    /// the given `handler_builder`. Finally, it calls the unary request interface of the
+    /// `RequestHandler` to process the request and produce a result.
+    // TODO: Convert to use async / await.
+    fn handle_unary_request_impl(
+        engine: E,
+        tracker: Tracker,
+        handler_builder: RequestHandlerBuilder<E::Snap>,
+    ) -> impl Future<Item = coppb::Response, Error = Error> {
+        // When this function is being executed, it may be queued for a long time, so that
+        // deadline may exceed.
+        future::result(tracker.req_ctx.deadline.check_if_exceeded())
+            .and_then(move |_| {
+                Self::async_snapshot(engine, &tracker.req_ctx.context)
+                    .map(|snapshot| (tracker, snapshot))
+            })
+            .and_then(move |(tracker, snapshot)| {
```

Comment:
why repeat `.and_then(move |(tracker, snapshot)| {` 3 times?

Reply:
Because tracker is moved to closure inside, so it has to be passed out, otherwise it will be consumed, which is not what we want. Notice that tracker is constructed before future is executed, so it must be moved inside closure, otherwise when it is referenced it has been destroyed (which is incorrect and of course not allowed by Rust)

Reply:
@breeswish 
How about making tracker and snapshot boxed? The size of `Future::Item` may be big currently.

Reply:
@breeswish some tests have been done, and u can see them in TIKV-1489.